### ISSUE TITLE
[FirebaseCore]: Wasm build crashes on launch in initializeApp

### ISSUE THREAD

**Issue by sma** [firebase/flutterfire#12623]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Which plugins are affected?

Core

### Which platforms are affected?

Web

### Description

I compiled an app to Wasm, but it crashes upon launch. I think, this is related to Firebase, but frankly, I'm a bit lost how to drill down the problem (luckily I found the `--no-strip-wasm` option), so I could at least get some debug symbols.

The relevant dependencies:
```
Dart SDK 3.5.0-36.0.dev
Flutter SDK 3.22.0-6.0.pre.25
├── cloud_firestore 4.16.1
├── firebase_core 2.29.0
```

The command to build the app:
```sh
flutter build web --wasm --no-strip-wasm
```

The crash:
```js
JavaScriptError
     at FirebaseCoreWeb.initializeApp inner (http://localhost:8080/main.dart.wasm:wasm-function[15504]:0x279d9e)
    at _awaitHelperWithTypeCheck closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 (http://localhost:8080/main.dart.wasm:wasm-function[2020]:0x14f963)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 trampoline (http://localhost:8080/main.dart.wasm:wasm-function[2025]:0x14f9ca)
    at _RootZone.runUnary (http://localhost:8080/main.dart.wasm:wasm-function[1228]:0x13afb3)
    at _Future._propagateToListeners (http://localhost:8080/main.dart.wasm:wasm-function[1207]:0x13ab5c)
    at _Future._completeWithValue (http://localhost:8080/main.dart.wasm:wasm-function[1204]:0x13a777)
    at _Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 (http://localhost:8080/main.dart.wasm:wasm-function[1199]:0x13a724)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 trampoline (http://localhost:8080/main.dart.wasm:wasm-function[1202]:0x13a74e)
```

I tried to make some sense of the of the WAT and I think
```wat
(func $FirebaseCoreWeb.initializeApp inner (;15504;) (param $var0 (ref $_AsyncSuspendState_190)) (param $var1 (ref null $#Top)) (param $var2 (ref null $#Top)) (param $var3 (ref null $Object)) (result (ref null $#Top))
```
is the method where the exception is thrown eventually:
```wat
      catch_all
        local.get $var0
        struct.get $_AsyncSuspendState_190 $field5
        call $_JavaScriptError
        call $StackTrace.current
        call $_Completer.completeError
        drop
        br $label8
      end
      unreachable
```

This might map [to this code](https://github.com/firebase/flutterfire/blob/213608cf7d889aeb60084ef5fb74a1e74178d7ba/packages/firebase_core/firebase_core_web/lib/src/interop/core.dart#L26) however, there's no exception handling and there's no mention of `JavaScriptError` in this project. So it might be automatically generated by the Dart compiler.

### Reproducing the issue

I reduced my app to 
```js
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);
}
```
and I still get the exception, so, it's definitively that call. There's no root cause, unfortunately.

### Firebase Core version

2.29.0

### Flutter Version

3.22.0

### Flutter dependencies

<details>
<summary>Expand <code>Flutter dependencies</code> snippet</summary>
<br>

```yaml

Replace this line with the contents of your `flutter pub deps -- --style=compact`.

```

</details>

### Additional context and comments

**Note:** The app works just fine without wasm, so it isn't a configuration error.

**Comment by Lyokone** [comment]
Thanks for the report, I'm having a look

**Comment by Lyokone** [comment]
I was able to compile (`flutter build web --wasm`) without any issues using the latest flutter versino

**Comment by sma** [comment]
Compiling isn't the issue. Running is.

**Comment by Lyokone** [comment]
@sma I'm able to run without issue as well using `serve` for instance, please provide a complete reproduction sample.

**Comment by sma** [comment]
Here's a complete example (with the Firebase options redacted):
```dart
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/material.dart';

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  const options = FirebaseOptions(
    apiKey: '<...>',
    appId: '<...>',
    messagingSenderId: '<...>',
    projectId: '<...>',
  );
  await Firebase.initializeApp(options: options);
  runApp(const MaterialApp(home: Scaffold(body: Text('It works!'))));
}
```

The example works just fine with
```
$ flutter --version
Flutter 3.22.0-9.0.pre.18 • channel master • https://github.com/flutter/flutter.git
Framework • revision 53cba24de3 (9 hours ago) • 2024-04-11 21:07:19 -0700
Engine • revision 6b37b17099
Tools • Dart 3.5.0 (build 3.5.0-47.0.dev) • DevTools 2.34.1
```

using `firebase_core 2.29.0`.

When compiling the file with 
```
flutter build web -t lib/bug.dart --wasm --no-strip-wasm
```
and using running the app locally with httpd as documented:
```
cd build/web
dhttpd '--headers=Cross-Origin-Embedder-Policy=credentialless;Cross-Origin-Opener-Policy=same-origin'
```

I get this crash in Chrome 123.0.6312.107:  
```
    at FirebaseCoreWeb.initializeApp inner (http://localhost:8080/main.dart.wasm:wasm-function[6839]:0x1175eb)
    at _awaitHelperWithTypeCheck closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 (http://localhost:8080/main.dart.wasm:wasm-function[1764]:0xa9d15)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 trampoline (http://localhost:8080/main.dart.wasm:wasm-function[1769]:0xa9d7c)
    at _RootZone.runUnary (http://localhost:8080/main.dart.wasm:wasm-function[1028]:0x96ce0)
    at _Future._propagateToListeners (http://localhost:8080/main.dart.wasm:wasm-function[1007]:0x9682c)
    at _Future._completeWithValue (http://localhost:8080/main.dart.wasm:wasm-function[1004]:0x96447)
    at _Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 (http://localhost:8080/main.dart.wasm:wasm-function[999]:0x963f4)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:721:29 trampoline (http://localhost:8080/main.dart.wasm:wasm-function[1002]:0x9641e)
```

I switch on "break on exceptions" in the chrome dev tools and got this, but I'm not sure it's related to the crash:
<img width="566" alt="image" src="https://github.com/firebase/flutterfire/assets/14989/7e7a2ee3-2d45-404e-bcab-cfe7d4664909">

As already mentioned, this crashes only with WASM, the "normal" JavaScript-Version works as expected, for example in Firefox 124.

I tested the sub expressions from the `if` and `getDefaultAppConfig()` returns `undefined`. 

This seems to call [this method](https://github.com/firebase/firebase-js-sdk/blob/7db2e04c6233df582c717d9cede9af2d445db6dd/packages/util/src/defaults.ts#L100) and somehow the configuration passed in Dart to that Dart method isn't transported to JS via WASM … I'd guess.

**Comment by sma** [comment]
If I should guess, I'd say that this code snippet from `Firebase_core_web.dart` is unable to catch the exception that occurs because there's no default configuration. The code tries to detect a default application before it tries into "manually" set the options.

```dart
    if (name == null || name == defaultFirebaseAppName) {
      bool defaultAppExists = false;

      try {
        app = firebase.app();
        defaultAppExists = true;
      } catch (e) {
        // noop
      }
```

Perhaps, that's a Dart compiler bug? IDK.

**Comment by sma** [comment]
Has been fixed in the Dart compiler.

**Comment by Knupper** [comment]
Was this fix part of the latest flutter release 3.22.0 ? I still have the same issue there.

**Comment by sma** [comment]
IDK, I'm on master (with Dart 3.5) where it has been fixed.
